pkg/util/xds/server.go

```diff
@@ -156,6 +158,31 @@ func (s *server) process(stream stream, reqCh <-chan *v2.DiscoveryRequest, defau
 		}
 	}
 
+	peekResponse := func(req cache.Request) *cache.Response {
+		versionInfo := req.VersionInfo
+		req.VersionInfo = "" // trick cache into returning a response despite unchanged version
+		resp, err := s.cache.Fetch(stream.Context(), req)
+		if err != nil || resp == nil {
+			return nil
+		}
+		resp.Request.VersionInfo = versionInfo // restore the original value
+		return resp
+	}
+	scheduleResponse := func(resp cache.Response) (value chan cache.Response, cancel func()) {
+		value = make(chan cache.Response, 1)
+		value <- resp
+		cancel = func() {
+			close(value)
+
+			// consume chanel
```

Comment:
why do we need to consume channel after closing it?

Reply:
after a buffered channel is closed, it can still be read as long as it has elements in the buffer (see https://play.golang.org/p/-Hdv_4AybAk)

Reply:
ok, thanks for the explanation